src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/ProjectDiagnosticSource.cs

```diff
@@ -11,28 +11,59 @@
 
 namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
 
-internal sealed record class ProjectDiagnosticSource(Project Project, Func<DiagnosticAnalyzer, bool>? ShouldIncludeAnalyzer) : IDiagnosticSource
+internal abstract class ProjectDiagnosticSource : IDiagnosticSource
 {
+    protected ProjectDiagnosticSource(Project project)
+    {
+        Project = project;
+    }
+
+    protected Project Project { get; }
+
+    public static ProjectDiagnosticSource CreateForFullSolutionAnalysisDiagnostics(Project project, Func<DiagnosticAnalyzer, bool>? shouldIncludeAnalyzer)
+        => new FullSolutionAnalysisDiagnosticSource(project, shouldIncludeAnalyzer);
+
+    public static ProjectDiagnosticSource CreateForCodeAnalysisDiagnostics(Project project, ICodeAnalysisDiagnosticAnalyzerService codeAnalysisService)
+        => new CodeAnalysisDiagnosticSource(project, codeAnalysisService);
```

Comment:
We now have distinct sources for FSA and Code Analysis based workspace diagnostics.